src/idl_gen_ts.cpp

```diff
@@ -953,6 +954,44 @@ class TsGenerator : public BaseGenerator {
     return ret;
   }
 
+  std::string GenUnionMapFunc(const Type &union_type, import_set &imports) {
+    if (union_type.enum_def) {
+      const auto &enum_def = *union_type.enum_def;
+      const auto enum_type = AddImport(imports, enum_def, enum_def).name;
+
+      auto ret = "\n\nexport type " + enum_type + "Mapping = {\n";
+
+      const auto union_enum_loop = [&](bool object_api) {
+        for (auto& it : enum_def.Vals()) {
+          const auto &ev = *it;
+          if (ev.IsZero()) { continue; }
+
+          const auto union_import =
+              AddImport(imports, enum_def, *ev.union_type.struct_def);
+          const auto union_name =
+              object_api ? union_import.object_name : union_import.name;
+
+          ret += "  [" + enum_type + "." + namer_.Variant(ev) + "]: " +
+                 union_name + ";\n";
+        }
+      };
+
+      union_enum_loop(false);
+
+      ret += "};";
+
+      if (parser_.opts.generate_object_based_api) {
+        ret += "\n\nexport type " + enum_type + "MappingT = {\n";
+        union_enum_loop(true);
```

Comment:
Maybe it would be better union_enum_loop lambda return a string and append to ret here, it is more readable